crm_compassion/models/weekly_demand.py

```diff
@@ -99,61 +113,101 @@ def _compute_demand_events(self):
+        sub_sponsored = self.env['recurring.contract'].search_count([
+            ('origin_id.type', '=', 'sub'),
+            ('start_date', '>=', fields.Date.to_string(start_date))
+        ])
+        return float(sub_sponsored) / STATS_DURATION
+
     @api.depends('number_children_website', 'number_children_ambassador',
-                 'number_children_events')
+                 'number_sub_sponsorship', 'number_children_events')
     @api.multi
     def _compute_demand_total(self):
         for week in self:
             week.total_demand = week.number_children_website + \
+                week.number_sub_sponsorship + \
                 week.number_children_ambassador + week.number_children_events
 
     @api.model
     def _default_unsponsored_web(self):
         """ Compute average of unsponsored children on the website since
         one year.
         """
-        start_date = datetime.today() - timedelta(weeks=52)
+        start_date = datetime.today() - timedelta(weeks=STATS_DURATION)
         web_sponsored = self.env['recurring.contract'].search_count([
             ('channel', '=', 'internet'),
             ('start_date', '>=', fields.Date.to_string(start_date))
         ])
         allocate_per_week = int(self.env['ir.config_parameter'].get_param(
             'crm_compassion.number_children_web'))
-        return allocate_per_week - (float(web_sponsored) / 52.0)
+        return allocate_per_week - (float(web_sponsored) / STATS_DURATION)
 
     @api.model
     def _default_unsponsored_ambassador(self):
         """ Compute average of unsponsored children from ambassadors since
         one year.
         """
-        start_date = datetime.today() - timedelta(weeks=52)
+        start_date = datetime.today() - timedelta(weeks=STATS_DURATION)
         ambass_sponsored = self.env['recurring.contract'].search_count([
             ('origin_id.type', '=', 'partner'),
             ('origin_id.partner_id', '!=', False),
             ('start_date', '>=', fields.Date.to_string(start_date))
         ])
         allocate_per_week = int(self.env['ir.config_parameter'].get_param(
             'crm_compassion.number_children_ambassador'))
-        return allocate_per_week - (float(ambass_sponsored) / 52.0)
+        return allocate_per_week - (float(ambass_sponsored) / STATS_DURATION)
+
+    @api.depends('number_sub_sponsorship')
+    @api.multi
+    def _compute_resupply_sub(self):
+        """ Compute SUB resupply. """
+        for week in self[:int(SUB_DURATION / 7)]:
+            rejected_sub = self.env['recurring.contract'].search_count([
+                ('origin_id.type', '=', 'sub'),
+                ('end_date', '<=', week.week_end_date),
+                ('end_date', '>=', week.week_start_date),
+                ('start_date', '>=', fields.Date.from_string(
+                    week.week_start_date) - timedelta(days=SUB_DURATION))
+            ])
+            week.resupply_sub = rejected_sub
```

Comment:
After checking, this is wrong and returns always 0, because week_end_date and week_start_date will always be in the future. So we never have any sponsorships with end_date in the future.

It would  be the correct number for the past but we cannot predict the future. So the correct computation is :

````
number of sub started 90 days ago * (sub_reject_average / sub_average)
```

@davidwul is this formula correct ? :laughing: